Here is another failure in TestKVNemesisMultiNode (not the LeaderLeases variant, which was the previous one) with some more details.

Failure:
```
kvnemesis.go:173: error applying x.AdminMerge(ctx, tk(302066519320058380)) // kv/kvserver/replica_command.go:993: merge failed: on 1 Subsume at /Table/100/"1ca4d4b8330920c4": injected error: kv/kvserver/replica_command.go:993: merge failed: on 1 Subsume at /Table/100/"1ca4d4b8330920c4": injected error
```

Logs:
```
I241212 14:11:15.760044 13944 kv/kvserver/replica_command.go:842 ⋮ [T1,Vsystem,n2,s2,r73/3:‹/Table/{71-100/"1ca4…}›,*kvpb.AdminMergeRequest] 1079  initiating a merge of r84:‹/Table/100/"{1ca4d4b8330920c4"-f47e2329d9f733df"}› [(n3,s3):1, (n4,s4):2, (n2,s2):3, next=4, gen=10, sticky=9223372036.854775807,2147483647] into this range (manual)
I241212 14:11:15.761165 13724 kv/kvnemesis/watcher.go:283 ⋮ [-] 1080  watcher reached frontier 1734012672.754360000,0 lagging by 3.006805s
I241212 14:11:15.761176 13724 kv/kvnemesis/watcher.go:283 ⋮ [-] 1081  watcher reached frontier 1734012672.756877000,0 lagging by 3.004298s
I241212 14:11:15.761519 2498 kv/kvnemesis/kvnemesis_test.go:126 ⋮ [-] 1082  inserting illegal lease index for /Table/100/‹"1ca4d4b8330920c4"› (seen 1 times)
I241212 14:11:15.761916 2497 kv/kvnemesis/kvnemesis_test.go:101 ⋮ [-] 1083  inserting reproposal error for /Table/100/‹"1ca4d4b8330920c4"› (seen 1 times)
I241212 14:11:15.761964 2497 kv/kvserver/replica_application_result.go:222 ⋮ [T1,Vsystem,n2,s2,r84/3:‹/Table/100/"{1ca4d4…-f47e23…}›,raft] 1084  failed to repropose 131d5bda712e2fbd at idx 24 with new lease index: on ‹1 Subsume› at /Table/100/‹"1ca4d4b8330920c4"›: injected error

```

The log statement in the code has a comment
```
				// Commonly the error is an inability to repropose the command due to
				// the closed timestamp having advanced past the write timestamp at
				// which the command was originally evaluated, so this path is hit
				// in practice.
```

So I am wondering whether this is an entirely reasonable failure, and we are just not ignoring it. There are a few other commands in the logs that encounter "failed to repropose". Looking at `validator.processOp`, it does ignore some errors, and I tried to see if other writes, like `PutOperation` are ignoring such an error via the call to `validator.checkError`, but couldn't figure it out. The `MergeOperation` error exceptions don't seem to include such reproposal errors, so I wonder whether they don't have these errors since all the other writes are in the context of the merge txn, and so occurring at a timestamp that is not closed. While this SubsumeRequest write is not part of the txn, so could have a timestamp that is already closed, if it needed to retry many times.

--- a/pkg/kv/kvpb/api.go
+++ b/pkg/kv/kvpb/api.go
@@ -2113,7 +2113,7 @@ func (r *RefreshRangeRequest) flags() flag {
 	return isRead | isTxn | isRange | updatesTSCache
 }
 
-func (*SubsumeRequest) flags() flag    { return isRead | isAlone | updatesTSCache }
+func (*SubsumeRequest) flags() flag    { return isWrite | isAlone | updatesTSCache }
 func (*RangeStatsRequest) flags() flag { return isRead }
 func (*QueryResolvedTimestampRequest) flags() flag {
 	return isRead | isRange | requiresClosedTSOlderThanStorageSnapshot

--- a/pkg/kv/kvserver/batcheval/cmd_subsume.go
+++ b/pkg/kv/kvserver/batcheval/cmd_subsume.go
@@ -71,6 +71,55 @@ func declareKeysSubsume(
 //
 // The period of time after intents have been placed but before the merge
 // transaction is complete is called the merge's "critical phase".
+//
+// SubsumeRequest does a write on newer cluster versions. This is accomplished
+// by (a) marking SubsumeRequest.flags() with isWrite, and (b) returning a
+// non-empty Result.Replicated. Both are necessary for SubsumeRequest to be
+// replicated via Raft. The write is limited to updating the
+// RangeForceFlushKey with a value corresponding to the index of the raft
+// entry of the Subsume. Since this write is to a replicated Range-ID local
+// key (and because it is done non-transactionally -- see
+// kvserver.AdminMerge), it will not be rolled back if the distributed
+// transaction doing the merge aborts. This is ok since the RangeForceFlushKey
+// only provides a directive to the replication flow control v2 (RACv2)
+// machinery to force-flush up to a particular Raft index. Additionally,
+// safety relies on SubsumeRequest only writing to a Range-ID local key:
+//
+//   - Neither the SubsumeResponse.{MVCCStats,RangeIDLocalMVCCStats} includes
+//     the effect of this write, which is ok since
+//     MVCCState-RangeIDLocalMVCCStats is added to the merged range's stats and
+//     the RHS's RangeForceFlushKey will be deleted.
+//
+//   - SubsumeResponse.LeaseAppliedIndex, say x, includes all requests preceding
+//     this SubsumeRequest, but not this request. Since the merge txn only waits
+//     for application on all RHS replicas up to x, it is possible that some
+//     replica has not applied the SubsumeRequest and gets merged with the LHS.
+//     This is ok since the RangeForceFlushKey of the RHS is deleted by the
+//     merge.
+//
+// The SubsumeResponse.{FreezeStart,ClosedTimestamp,ReadSummary} are used in
+// varying degrees to adjust the timestamp cache of the merged range (see
+// Store.MergeRange). On the surface, this behavior is unaffected by whether
+// SubsumeRequest is a read or write, since the effect of this SubsumeRequest
+// on the timestamp cache is ignored either way (this request evaluation
+// updates the timestamp cache after the call to Subsume returns), and is not
+// relevant for correctness. However, there is a hazard if this SubsumeRequest
+// (when it is a write) bumps the closed timestamp, say from t1 to t2.
+// Followers could subsequently admit reads (on the RHS, before the merge)
+// with timestamps in the interval (t1, t2], that will not be accounted for in
+// the timestamp cache of the merged range (since the closed timestamp
+// returned in SubsumeResponse.ClosedTimestamp=t1). Note that such follower
+// writes are possible because followers do not freeze the range like the
+// leaseholder does. The simple way to avoid this hazard is to not bump the
+// closed timestamp when proposing a SubsumeRequest. This is accomplished by
+// two means:
+//   - Side transport (see Replica.BumpSideTransportClosed): Does not bump the
+//     closed timestamp if the range merge is in progress. And Subsume ensures
+//     that the merge is in progress by calling WatchForMerge below.
+//   - Closed timestamp replicated via Raft (see
+//     propBuf.allocateLAIAndClosedTimestampLocked): The closed timestamp is not
+//     advanced if the BatchRequest contains a single request which is a
+//     SubsumeRequest.
 func Subsume(
 	ctx context.Context, readWriter storage.ReadWriter, cArgs CommandArgs, resp kvpb.Response,
 ) (result.Result, error) {

--- a/pkg/kv/kvserver/flow_control_integration_test.go
+++ b/pkg/kv/kvserver/flow_control_integration_test.go
@@ -5774,9 +5774,10 @@ func TestFlowControlSendQueueRangeSplitMerge(t *testing.T) {
 -- We will exhaust the tokens across all streams while admission is blocked on
 -- n3, using a single 4 MiB (deduction, the write itself is small) write. Then,
 -- we will write a 1 MiB put to the range, split it, write a 1 MiB put to the
--- LHS range, merge the ranges, and write a 1 MiB put to the merged range. We
--- expect that at each stage where a send queue develops n1->s3, the send queue
--- will be flushed by the range merge and range split range operations.`)
+-- LHS range, and a 1MiB put to the RHS range, merge the ranges, and write a 1
+-- MiB put to the merged range. We expect that at each stage where a send
+-- queue develops n1->s3, the send queue will be flushed by the range merge
+-- and range split range operations.`)
 	h.comment(`
 -- Start by exhausting the tokens from n1->s3 and blocking admission on s3.
 -- (Issuing 4x1MiB regular, 3x replicated write that's not admitted on s3.)`)
@@ -5838,13 +5839,11 @@ ORDER BY streams DESC;
 	h.waitForAllTokensReturnedForStreamsV2(ctx, 0, /* serverIdx */
 		testingMkFlowStream(0), testingMkFlowStream(1))
 
-	// TODO(kvoli): Uncomment once we resolve the subsume wait for application
-	// issue. See #136649.
-	// h.comment(`(Sending 1 MiB put request to post-split RHS range)`)
-	// h.put(ctx, roachpb.Key(right.StartKey), 1, admissionpb.NormalPri)
-	// h.comment(`(Sent 1 MiB put request to post-split RHS range)`)
-	// h.waitForAllTokensReturnedForStreamsV2(ctx, 0, /* serverIdx */
-	// 	testingMkFlowStream(0), testingMkFlowStream(1))
+	h.comment(`(Sending 1 MiB put request to post-split RHS range)`)
+	h.put(ctx, roachpb.Key(right.StartKey), 1, admissionpb.NormalPri)
+	h.comment(`(Sent 1 MiB put request to post-split RHS range)`)
+	h.waitForAllTokensReturnedForStreamsV2(ctx, 0, /* serverIdx */
+		testingMkFlowStream(0), testingMkFlowStream(1))
 
 	h.comment(`
 -- Send queue and flow token metrics from n1, post-split and 1 MiB put on

--- a/pkg/kv/kvserver/replica_command.go
+++ b/pkg/kv/kvserver/replica_command.go
@@ -881,7 +881,39 @@ func (r *Replica) AdminMerge(
 		// Intents have been placed, so the merge is now in its critical phase. Get
 		// a consistent view of the data from the right-hand range. If the merge
 		// commits, we'll write this data to the left-hand range in the merge
-		// trigger.
+		// trigger. This consistent view is dependent on two things (a) the
+		// SubsumeRequest is evaluated at a leaseholder that is at least as recent
+		// as the leaseholder that evaluated the deletion intent placed on the RHS
+		// RangeDescriptor, (b) the SubsumeRequest freezes the range for new reads
+		// and writes at the leaseholder, and freezes the closed timestamp so
+		// follower reads cannot occur at a timestamp beyond what is accounted for
+		// in SubsumeResponse.ClosedTimestamp (see batcheval.Subsume for more
+		// details).
+		//
+		// When SubsumeRequest does a write (on newer cluster versions), (a) is
+		// guaranteed by the fact that this request needs to be replicated and so
+		// cannot be successfully processed by a stale leaseholder (due to the
+		// protection in RaftCommand.ProposerLeaseSequence). When SubsumeRequest
+		// is a read, it may seem that (a) is not guaranteed since the request
+		// below is sent outside the txn, and does not set a timestamp, so could
+		// be assigned a timestamp lower than the txn timestamp, and routed to an
+		// older leaseholder. The reason this doesn't happen is subtle:
+		// - Say the txn (whose txn coordinator is this node) got assigned a
+		//   timestamp t1.
+		// - The intent put went to the RHS leaseholder which was ahead, at
+		//   timestamp t2. The response to the intent put will bump up the local
+		//   hlc.Timestamp to t2.
+		// - This SubsumeRequest does not set kvpb.Header.Timestamp, but it will
+		//   include in kvpb.Header.Now a value that is >= t2. When this request
+		//   is received by an old leaseholder (the aforementioned hazard), the
+		//   old leaseholder will bump its hlc.Clock to >= t2 and stop being the
+		//   leaseholder, and reject the request. The request will get eventually
+		//   (successfully) retried at a leaseholder that has the lease at
+		//   timestamp >= t2.
+		//
+		// This must be a single request in a BatchRequest: there are multiple
+		// places that do special logic (needed for safety) that rely on
+		// BatchRequest.IsSingleSubsumeRequest() returning true.
 		br, pErr := kv.SendWrapped(ctx, r.store.DB().NonTransactionalSender(),
 			&kvpb.SubsumeRequest{
 				RequestHeader: kvpb.RequestHeader{Key: rightDesc.StartKey.AsRawKey()},
@@ -958,7 +990,7 @@ func (r *Replica) AdminMerge(
 		}
 		if !errors.HasType(err, (*kvpb.TransactionRetryWithProtoRefreshError)(nil)) {
 			if err != nil {
-				return reply, kvpb.NewErrorf("merge failed: %s", err)
+				return reply, kvpb.NewError(errors.Wrap(err, "merge failed"))
 			}
 			return reply, nil
 		}

--- a/pkg/kv/kvserver/replica_proposal_buf.go
+++ b/pkg/kv/kvserver/replica_proposal_buf.go
@@ -732,6 +732,7 @@ func (b *propBuf) allocateLAIAndClosedTimestampLocked(
 	// applies. The command application side doesn't bother protecting against
 	// such regressions. Besides, the considerations for brand new leases below
 	// also apply.
+	//
 	// - For a brand new lease, one might think that the lease start time can be
 	// considered a closed timestamp(*) since, if this replica gets the lease, it
 	// will not evaluate writes at lower timestamps. Unfortunately, there's a
@@ -740,14 +741,24 @@ func (b *propBuf) allocateLAIAndClosedTimestampLocked(
 	// happen that the range is in the process of merging with its left neighbor.
 	// If this range has already been Subsumed as the RHS of a merge then, after
 	// merge, the joint range will allow writes to the former RHS's key space at
-	// timestamps above the RHS's freeze start (which is below the start time of
-	// this lease). Thus, if this lease were to close its start timestamp while
-	// subsumed, then it'd be possible for follower reads to be served before the
-	// merge finalizes at timestamps that would become un-closed after the merge.
+	// timestamps above the RHS's closed timestamp (say t1) at freeze start
+	// (which is below the start time of this lease, say t2). Thus, if this
+	// lease were to close its start timestamp (t2) while subsumed, then it'd be
+	// possible for follower reads to be served before the merge finalizes at
+	// timestamps that would become un-closed after the merge, i.e., interval
+	// (t1, t2].
+	//
 	// Since this scenario is about subsumed ranges, we could make a distinction
 	// between brand new leases for subsumed ranges versus other brand new leases,
 	// and let the former category close the lease start time. But, for
 	// simplicity, we don't close timestamps on any lease requests.
+	//
+	// The same logic applies to SubsumeRequest (when it performs a write): the
+	// closed timestamp it gathers in SubsumeResponse.ClosedTimestamp is t1, but
+	// if the proposal of the SubsumeRequest advances the closed timestamp to
+	// t2, reads would be allowed over the interval (t1, t2] which would
+	// subsequently become un-closed. So SubsumeRequests also don't advance the
+	// closed timestamp.
 	//
 	// As opposed to lease requests, lease transfers behave like regular
 	// proposals: they get a closed timestamp based on closedTSTarget. Note that
@@ -763,7 +774,7 @@ func (b *propBuf) allocateLAIAndClosedTimestampLocked(
 	// timestamps carried by lease requests, make sure to resurrect the old
 	// TestRejectedLeaseDoesntDictateClosedTimestamp and protect against that
 	// scenario.
-	if p.Request.IsSingleRequestLeaseRequest() {
+	if p.Request.IsSingleRequestLeaseRequest() || p.Request.IsSingleSubsumeRequest() {
 		return lai, hlc.Timestamp{}, nil
 	}
 

--- a/pkg/kv/kvserver/testdata/flow_control_integration_v2/send_queue_range_split_merge
+++ b/pkg/kv/kvserver/testdata/flow_control_integration_v2/send_queue_range_split_merge
@@ -4,9 +4,10 @@ echo
 -- We will exhaust the tokens across all streams while admission is blocked on
 -- n3, using a single 4 MiB (deduction, the write itself is small) write. Then,
 -- we will write a 1 MiB put to the range, split it, write a 1 MiB put to the
--- LHS range, merge the ranges, and write a 1 MiB put to the merged range. We
--- expect that at each stage where a send queue develops n1->s3, the send queue
--- will be flushed by the range merge and range split range operations.
+-- LHS range, and a 1MiB put to the RHS range, merge the ranges, and write a 1
+-- MiB put to the merged range. We expect that at each stage where a send
+-- queue develops n1->s3, the send queue will be flushed by the range merge
+-- and range split range operations.
 
 
 -- Start by exhausting the tokens from n1->s3 and blocking admission on s3.
@@ -115,6 +116,12 @@ SELECT store_id,
 (Sent 1 MiB put request to post-split LHS range)
 
 
+(Sending 1 MiB put request to post-split RHS range)
+
+
+(Sent 1 MiB put request to post-split RHS range)
+
+
 -- Send queue and flow token metrics from n1, post-split and 1 MiB put on
 -- each side.
 SELECT name, crdb_internal.humanize_bytes(value::INT8)
@@ -123,7 +130,7 @@ SELECT name, crdb_internal.humanize_bytes(value::INT8)
      AND name != 'kvflowcontrol.send_queue.count'
 ORDER BY name ASC;
 
-  kvflowcontrol.send_queue.bytes                                    | 1.0 MiB  
+  kvflowcontrol.send_queue.bytes                                    | 2.0 MiB  
   kvflowcontrol.send_queue.prevent.count                            | 0 B      
   kvflowcontrol.send_queue.scheduled.deducted_bytes                 | 0 B      
   kvflowcontrol.send_queue.scheduled.force_flush                    | 0 B      
@@ -142,7 +149,7 @@ SELECT store_id,
 -----------+------------------------+------------------------+------------------------+-------------------------
   1        | 4.0 MiB                | 2.0 MiB                | 4.0 MiB                | 2.0 MiB                 
   2        | 4.0 MiB                | 2.0 MiB                | 4.0 MiB                | 2.0 MiB                 
-  3        | 0 B                    | -4.0 MiB               | 0 B                    | -3.0 MiB                
+  3        | 0 B                    | -5.0 MiB               | 0 B                    | -3.0 MiB                
 
 
 -- (Merging ranges.)
@@ -160,7 +167,7 @@ ORDER BY name ASC;
   kvflowcontrol.send_queue.prevent.count                            | 0 B      
   kvflowcontrol.send_queue.scheduled.deducted_bytes                 | 0 B      
   kvflowcontrol.send_queue.scheduled.force_flush                    | 0 B      
-  kvflowcontrol.tokens.send.elastic.deducted.force_flush_send_queue | 2.0 MiB  
+  kvflowcontrol.tokens.send.elastic.deducted.force_flush_send_queue | 3.0 MiB  
   kvflowcontrol.tokens.send.elastic.deducted.prevent_send_queue     | 0 B      
   kvflowcontrol.tokens.send.regular.deducted.prevent_send_queue     | 0 B      
 SELECT store_id,
@@ -196,7 +203,7 @@ ORDER BY name ASC;
   kvflowcontrol.send_queue.prevent.count                            | 0 B      
   kvflowcontrol.send_queue.scheduled.deducted_bytes                 | 0 B      
   kvflowcontrol.send_queue.scheduled.force_flush                    | 0 B      
-  kvflowcontrol.tokens.send.elastic.deducted.force_flush_send_queue | 2.0 MiB  
+  kvflowcontrol.tokens.send.elastic.deducted.force_flush_send_queue | 3.0 MiB  
   kvflowcontrol.tokens.send.elastic.deducted.prevent_send_queue     | 0 B      
   kvflowcontrol.tokens.send.regular.deducted.prevent_send_queue     | 0 B      
 SELECT store_id,
@@ -228,7 +235,7 @@ ORDER BY name ASC;
   kvflowcontrol.send_queue.prevent.count                            | 0 B      
   kvflowcontrol.send_queue.scheduled.deducted_bytes                 | 0 B      
   kvflowcontrol.send_queue.scheduled.force_flush                    | 0 B      
-  kvflowcontrol.tokens.send.elastic.deducted.force_flush_send_queue | 2.0 MiB  
+  kvflowcontrol.tokens.send.elastic.deducted.force_flush_send_queue | 3.0 MiB  
   kvflowcontrol.tokens.send.elastic.deducted.prevent_send_queue     | 0 B      
   kvflowcontrol.tokens.send.regular.deducted.prevent_send_queue     | 0 B      
 SELECT store_id,

